packages/ckb-sdk-core/src/index.ts

```diff
@@ -218,8 +262,201 @@ class Core {
+    if (tx.txStatus.status !== 'committed') throw new Error('Transaction is not committed yet')
+
+    const depositBlockHeader = await this.rpc.getBlock(tx.txStatus.blockHash).then(b => b.header)
+    const encodedBlockNumber = this.utils.toHexInLittleEndian(depositBlockHeader.number, 8)
+
+    const fromAddress = this.utils.bech32Address(cellStatus.cell.output.lock.args)
+
+    const rawTx = await this.generateRawTransaction({
+      fromAddress,
+      toAddress: fromAddress,
+      capacity: '0x0',
+      fee,
+      safeMode: true,
+      deps: this.config.secp256k1Dep!,
+      capacityThreshold: BigInt(0),
+      cells,
+    })
+
+    rawTx.outputs.splice(0, 1)
+    rawTx.outputsData.splice(0, 1)
+
+    rawTx.inputs.unshift({ previousOutput: outPoint, since: '0x0' })
+    rawTx.outputs.unshift(tx.transaction.outputs[+outPoint.index])
+    rawTx.cellDeps.push({ outPoint: this.config.daoDep!.outPoint, depType: 'code' })
+    rawTx.headerDeps.push(depositBlockHeader.hash)
+    rawTx.outputsData.unshift(encodedBlockNumber)
+    rawTx.witnesses.unshift({
+      lock: '',
+      inputType: '',
+      outputType: '',
+    })
+    return rawTx
+  }
+
+  public generateDaoWithdrawTransaction = async ({
+    depositOutPoint,
+    withdrawOutPoint,
+    fee,
+  }: {
+    depositOutPoint: CKBComponents.OutPoint
+    withdrawOutPoint: CKBComponents.OutPoint
+    fee: bigint | string
+  }): Promise<CKBComponents.RawTransactionToSign> => {
+    if (!this.config.secp256k1Dep) {
+      await this.loadSecp256k1Dep()
+    }
+    if (!this.config.daoDep) {
+      await this.loadDaoDep()
+    }
+
+    const DAO_LOCK_PERIOD_EPOCHS = 180
+    const cellStatus = await this.rpc.getLiveCell(withdrawOutPoint, true)
+    if (cellStatus.status !== 'live') throw new Error('Cell is not live yet')
+
+    const tx = await this.rpc.getTransaction(withdrawOutPoint.txHash)
+    if (tx.txStatus.status !== 'committed') throw new Error('Transaction is not committed yet')
+
+    /* eslint-disable */
+    const depositBlockNumber = +this.utils.bytesToHex(this.utils.hexToBytes(cellStatus.cell.data?.content ?? '').reverse())
+    /* eslint-enable */
+
+    const depositBlockHeader = await this.rpc.getBlockByNumber(BigInt(depositBlockNumber)).then(block => block.header)
+    const depositEpoch = this.utils.parseEpoch(depositBlockHeader.epoch)
+
+    const withdrawBlockHeader = await this.rpc.getBlock(tx.txStatus.blockHash).then(block => block.header)
+    const withdrawEpoch = this.utils.parseEpoch(withdrawBlockHeader.epoch)
+
+    const withdrawFraction = withdrawEpoch.index * withdrawEpoch.length
+    const depositFraction = depositEpoch.index * depositEpoch.length
+    let depositedEpochs = withdrawEpoch.number - depositEpoch.number
+    if (withdrawFraction > depositFraction) {
+      depositedEpochs += 1
+    }
+
+    const lockEpochs = Math.floor((depositedEpochs + DAO_LOCK_PERIOD_EPOCHS - 1) / 10) * 10
+    const minimalSinceEpochNumber = depositEpoch.number + lockEpochs
```

Comment:
> For a Nervos DAO withdrawing cell, the since field in the cell input should reflect the Nervos DAO cell’s lock period requirement, which is 180 epoches. For example, if one deposits into Nervos DAO at epoch 5, he/she can only expect to withdraw Nervos DAO at epoch 185, 365, 545, etc. Notice the calculation of lock period is independent of the calculation of interest. It’s totally valid to deposit at epoch 5, use a withdraw block at epoch 100, and use a since field at 185. 


I think `minimalSinceEpochNumber` can be just `depositBlockHeader.epoch.number + 180`. In fact, I dont quite understand this part L400-408.

Reply:
I've updated these part into https://github.com/nervosnetwork/ckb-sdk-js/pull/382/commits/5580419c387f5202d71045a027ff7505d3bbdd09

```
    // get the process of withdraw and deposit tx in their epoch
    const withdrawFraction = BigInt(withdrawEpoch.index) * BigInt(withdrawEpoch.length)
    const depositFraction = BigInt(depositEpoch.index) * BigInt(depositEpoch.length)
    // get the epochs between deposit and withdraw txs
    let depositedEpochs = BigInt(withdrawEpoch.number) - BigInt(depositEpoch.number)
    // if the process of withdraw in its epoch exceeds that of deposit, the deposited epochs should + 1
    if (withdrawFraction > depositFraction) {
      depositedEpochs += BigInt(1)
    }
    // calculate the lock epochs conform to 180 * n, for example
    // if the deposited epochs is 6, the lock epochs should be 180
    // if the deposited epochs is 330, the lock epochs should be 360 
    const lockEpochs = (depositedEpochs + BigInt(DAO_LOCK_PERIOD_EPOCHS - 1)) /
      BigInt(DAO_LOCK_PERIOD_EPOCHS) *
      BigInt(DAO_LOCK_PERIOD_EPOCHS)
    // compose the minimal
    const minimalSince = this.absoluteEpochSince(
      depositEpoch.number + lockEpochs,
      depositEpoch.index,
      depositEpoch.length
    )
```